zcash_primitives/src/sapling/zip32.rs

```diff
@@ -231,12 +229,34 @@ impl DiversifierKey {
     }
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum KeyIndex {
+    Master,
+    Child(ChildIndex),
+}
+
+impl KeyIndex {
+    fn new(depth: u8, i: u32) -> Option<Self> {
+        match i {
+            0 if depth == 0 => Some(KeyIndex::Master),
+            _ => ChildIndex::from_index(i).map(KeyIndex::Child),
```

Comment:
This allows non-zero indices at depth 0, which I think is invalid. (If it were not invalid then the child key at depth 0 would be ambiguous with the master key.)